File: src/test/java/com/polarbookshop/orderservice/order/domain/OrderJsonTest.java
package com.polarbookshop.orderservice.order.domain;

import static org.assertj.core.api.Assertions.assertThat;

import java.io.IOException;
import java.time.Instant;
import org.junit.jupiter.api.Test;
import org.springframework.boot.test.autoconfigure.json.JsonTest;
import org.springframework.boot.test.json.JacksonTester;
import org.springframework.boot.test.json.JsonContent;

@JsonTest
record OrderJsonTest(JacksonTester<Order> json) {

  @Test
  void serialize() throws IOException {
    Order order =
        Order.builder()
            .id(394L)
            .bookIsbn("1234567890")
            .bookName("Book Name")
            .bookPrice(9.90)
            .quantity(1)
            .status(OrderStatus.ACCEPTED)
            .createdDate(Instant.now())
            .lastModifiedDate(Instant.now())
            .version(21)
            .build();

    JsonContent<Order> jsonContent = json.write(order);

    assertThat(jsonContent).extractingJsonPathNumberValue("@.id").isEqualTo(order.id().intValue());
    assertThat(jsonContent).extractingJsonPathStringValue("@.bookIsbn").isEqualTo(order.bookIsbn());
    assertThat(jsonContent).extractingJsonPathStringValue("@.bookName").isEqualTo(order.bookName());
    assertThat(jsonContent)
        .extractingJsonPathNumberValue("@.bookPrice")
        .isEqualTo(order.bookPrice());
    assertThat(jsonContent).extractingJsonPathNumberValue("@.quantity").isEqualTo(order.quantity());
    assertThat(jsonContent)
        .extractingJsonPathStringValue("@.status")
        .isEqualTo(order.status().toString());
    assertThat(jsonContent)
        .extractingJsonPathStringValue("@.createdDate")
        .isEqualTo(order.createdDate().toString());
    assertThat(jsonContent)
        .extractingJsonPathStringValue("@.lastModifiedDate")
        .isEqualTo(order.lastModifiedDate().toString());
    assertThat(jsonContent).extractingJsonPathNumberValue("@.version").isEqualTo(order.version());
  }
}

File: src/test/kotlin/com/polarbookshop/orderservice/order/web/OrderJsonTest.kt
package com.polarbookshop.orderservice.order.web

import com.polarbookshop.orderservice.order.domain.Order
import com.polarbookshop.orderservice.order.domain.OrderStatus
import org.assertj.core.api.Assertions.assertThat
import org.junit.jupiter.api.Test
import org.springframework.boot.test.autoconfigure.json.JsonTest
import org.springframework.boot.test.json.JacksonTester
import java.time.Instant

@JsonTest
class OrderJsonTest(
    val json: JacksonTester<Order>,
) {
    @Test
    fun serialize() {
        val order =
            Order(
                id = 394L,
                bookIsbn = "1234567890",
                bookName = "Book Name",
                bookPrice = 9.90,
                quantity = 1,
                status = OrderStatus.ACCEPTED,
                createdDate = Instant.now(),
                lastModifiedDate = Instant.now(),
                version = 21,
            )
        val jsonContent = json.write(order)
        assertThat(jsonContent).extractingJsonPathNumberValue("@.id").isEqualTo(order.id.toInt())
        assertThat(jsonContent)
            .extractingJsonPathStringValue("@.bookIsbn")
            .isEqualTo(order.bookIsbn)
        assertThat(jsonContent)
            .extractingJsonPathStringValue("@.bookName")
            .isEqualTo(order.bookName)
        assertThat(jsonContent)
            .extractingJsonPathNumberValue("@.bookPrice")
            .isEqualTo(order.bookPrice)
        assertThat(jsonContent)
            .extractingJsonPathNumberValue("@.quantity")
            .isEqualTo(order.quantity)
        assertThat(jsonContent)
            .extractingJsonPathStringValue("@.status")
            .isEqualTo(order.status.toString())
        assertThat(jsonContent)
            .extractingJsonPathStringValue("@.createdDate")
            .isEqualTo(order.createdDate.toString())
        assertThat(jsonContent)
            .extractingJsonPathStringValue("@.lastModifiedDate")
            .isEqualTo(order.lastModifiedDate.toString())
        assertThat(jsonContent)
            .extractingJsonPathNumberValue("@.version")
            .isEqualTo(order.version)
    }
}
